src/file_system/error.rs

```diff
@@ -3,47 +3,43 @@
 //! These errors occur due to [`Label`] and [`Path`] parsing when using their respective `TryFrom`
 //! instances.
 
+use thiserror::Error;
+
 pub(crate) const EMPTY_PATH: Error = Error::Path(Path::Empty);
 
 pub(crate) const INVALID_UTF8: Error = Error::Label(Label::InvalidUTF8);
 pub(crate) const EMPTY_LABEL: Error = Error::Label(Label::Empty);
 pub(crate) const CONTAINS_SLASH: Error = Error::Label(Label::ContainsSlash);
 
 /// Error type for all file system errors that can occur.
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, Error)]
 pub enum Error {
     /// A `Label` specific error for parsing a [`Label`].
-    Label(Label),
+    #[error("Label error: {0}")]
+    Label(#[from] Label),
     /// A `Path` specific error for parsing a [`Path`].
-    Path(Path),
-}
-
-impl From<Label> for Error {
-    fn from(err: Label) -> Self {
-        Error::Label(err)
-    }
-}
-
-impl From<Path> for Error {
-    fn from(err: Path) -> Self {
-        Error::Path(err)
-    }
+    #[error("Path error: {0}")]
+    Path(#[from] Path),
 }
 
 /// Parse errors for when parsing a string to a [`Path`].
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, Error)]
 pub enum Path {
     /// An error signifying that a [`Path`] is empty.
+    #[error("Path is empty")]
     Empty,
 }
 
 /// Parse errors for when parsing a string to a [`Label`].
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq, Error)]
 pub enum Label {
     /// An error signifying that a [`Label`] is contains invalid UTF-8.
+    #[error("Label contains invalid UTF-8")]
     InvalidUTF8,
```

Comment:
Something I think we should do here is to actually have a `Label` inside `InvalidUTF8` an `ContainsSlash` so that we can provide more detail as to which `Label` was wrong making our errors more informative :) Could you add this while we're here editing this code?

Reply:
Great idea, editing this now.